extensions/undertow/deployment/src/main/java/io/quarkus/undertow/deployment/UndertowBuildStep.java

```diff
@@ -239,6 +246,49 @@ public void kubernetes(HttpConfig config, BuildProducer<KubernetesPortBuildItem>
         portProducer.produce(new KubernetesPortBuildItem(config.port, "http"));
     }
 
+    /**
+     * look for Servlet container initializers
+     *
+     */
+    @BuildStep(applicationArchiveMarkers = SERVLET_CONTAINER_INITIALIZER)
+    public List<ServletContainerInitializerBuildItem> servletContainerInitializer(
+            ApplicationArchivesBuildItem archives,
+            CombinedIndexBuildItem combinedIndexBuildItem,
+            BuildProducer<AdditionalBeanBuildItem> beans) throws IOException {
+        List<ServletContainerInitializerBuildItem> ret = new ArrayList<>();
+        Set<String> initializers = new HashSet<>();
+        for (ApplicationArchive i : archives.getAllApplicationArchives()) {
+            Path file = i.getChildPath(SERVLET_CONTAINER_INITIALIZER);
+            if (file != null) {
+                initializers.addAll(Files.readAllLines(file));
```

Comment:
I think you should use `ServiceUtil` here.

You could have comments in the file and `ServiceUtil` manages all of that (and is used everywhere else).

Reply:
I was not aware of ServiceUtil, however I am not sure that its current usage is safe.

In particular the use of getClass().getClassLoader() will return a class loader that has all the deployment related artifacts on the class path. As long as none of these contain the services this is fine, however if a deployment artifact or its dependency contains this service then we will have problems. ServletContainerInitializer's are fairly widely used, so it is possible this will cause problems down the road. 

This needs a bit of thought as to how to handle it properly.

Reply:
Yes, for now, the classloader situation is far from being perfect. I know @dmlloyd has a plan for that.

Not totally sure if we should try to find a solution now or just collect the issues we currently have to nourish this reflection.

Reply:
I don't think getting services from the deployment class path is a good idea.  We talked about the possibility having a synthetic class loader which represents the application but that would not be for situations like this.  Using `ServiceUtil` would not be right for this situation.

Reply:
Why would a synthetic ClassLoader not be right for this case? You could just pass it into ServiceUtil and it should limit services to the application

Reply:
You don't *need* a `ClassLoader` for this though, because it is a build step so you could scan the various application roots instead.

That said, if we were to introduce a new class loader for this (I was thinking we want to expose the three or so of them by build items) then there are a few other different useful applications for it.